_posts/2017/2017-12-31-avoid-double-negatives.md

```diff
@@ -0,0 +1,62 @@
+---
+title: 'Avoid Double Negatives'
+description: 'if unless else end'
+layout: article
+category: ruby
+image:
+  base: '2017/avoid-double-negatives'
+  alt: 'Two directions sign'
+  source: 'https://unsplash.com/photos/lPQIndZz8Mo'
+  credit: 'Pablo Garcia Saldaña'
+---
+
+Ruby has a bunch of delightful syntactic sugar. One example of this is `unless`, which you can use in place of an `if` with a negative.
+
+However given the flexibility of Ruby's syntax it is easy to make code harder to understand than it needs to be.
+
+## Instead of…
+
+...overcomplicating your conditionals.
+
+```ruby
+# Example 1
+unless something.nil?
+  # do something
+end
+
+# Example 2
+if !something.nil?
+  # do something
+end
+
+# Example 3
+if !!something
+  # do something
+end
+```
+
+
+## Use…
+
+```ruby
+# Instead of Examples 1,2 & 3
+if something
+  # do something
+end
+```
+
+
+## But why?
+
+Ruby’s conditional syntax is ‘truthy’. In Ruby’s case that means `nil` is equivalent to `false`. There is no need to do a negated `#nil?` check on an object as it is redundant as shown in the first two examples.
+
+The syntax of `!!` in the third example converts a value to boolean, but it is unnecessary and likely obscures your intention. If you _really_ want to do a `nil` check, you should use `#nil?` instead.
+
+You should definitely avoid using an `else` block with an `unless`: the `else` block is a double negative of the condition. Instead, switch the `unless` for an logically equivalent `if` and swap the code in the successful conditional block for the code in the `else` block. This might require some refactoring to the condition to make the resulting code easy to understand.
```

Comment:
This paragraph is hard to follow.

Can you use an code example (just for me) to illustrate what you're attempting to say?

Reply:
Reasoning about 

```ruby
unless condition
  # ...
else
  # the conditions that lead to this code
end
```

is super hard on your brain because it's logically 'not, not' the condition.

Reply:
Okay, maybe you want to say something like:

You should definitely avoid using `unless` and `else` in conjunction in a conditional. This structure is particularly hard to reason about because the `else` block acts as a double negative in the conditional. Instead, replace the `unless` for a logically equivalent `if` and swap the code between the two branches of the conditional. (Probably a better way to say this, particularly the last bit..)